sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+burnchain height `block-height`.
+
+* The `pox-addrs` property returns a tuple with two items: a list of up to two PoX addresses that received a PoX payout at that block height, and the amount of burnchain
+tokens paid to each address (note that per the blockchain consensus rules, each PoX payout will be the same for each address in the block-commit transaction).
+The list will include burn addresses -- that is, the unspendable addresses that miners pay to when there are no PoX addresses left to be paid.  During the prepare phase,
+there will be exactly one burn address reported. During the reward phase, up to two burn addresses may be reported in the event that some PoX reward slots are not claimed.
+
+The `addrs` list contains the same PoX address values passed into the PoX smart contract:
+   * They each have type signature `(tuple (hashbytes (buff 32)) (version (buff 1)))`
+   * The `version` field can be any of the following:
+      * `0x00` means this is a pay-to-public-key-hash (p2pkh) address, and `hashbytes` is the 20-byte hash160 of a single public key
+      * `0x01` means this is a pay-to-script-hash (p2sh) address, and `hashbytes` is the 20-byte hash160 of a redeemScript script
+      * `0x02` means this is a pay-to-witness-public-key-hash-over-pay-to-script-hash (p2wpkh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wpkh witness script
+      * `0x03` means this is a pay-to-witness-script-hash-over-pay-to-script-hash (p2wsh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wsh witness script
+      * `0x04` means this is a pay-to-witness-public-key-hash (p2wpkh) address, and `hashbytes` is the 20-byte hash160 of the witness script
+      * `0x05` means this is a pay-to-witness-script-hash (p2wsh) address, and `hashbytes` is the 32-byte sha256 of the witness script
+      * `0x06` means this is a pay-to-taproot (p2tr) address, and `hashbytes` is the 32-byte sha256 of the witness script
+
+**Rationale:**  This method empowers developers to query Bitcoin state.  Stacks
+2.1 adds the ability for smart contracts to query PoX payouts directly, with applications
+for Stacking pools and other Stacking-centric programs.
+
+**Examples:**
+
+```clarity
+(get-burn-block-info? header-hash u677050) ;; Returns (some 0xe67141016c88a7f1203eca0b4312f2ed141531f59303a1c267d7d83ab6b977d8)
+(get-burn-block-info? pox-addrs u677050) ;; Returns (some (tuple (addrs ((tuple (hashbytes 0x395f3643cea07ec4eec73b4d9a973dcce56b9bf1) (version 0x00)) (tuple (hashbytes 0x7c6775e20e3e938d2d7e9d79ac310108ba501ddb) (version 0x01)))) (payout u123)))
+```
+
+### New method: `slice`
+
+* **Input Signature:** `(slice (sequence sequence_A) (left-position uint) (right-position uint))`
+* **Output Signature:** `(optional sequence_A)`
+
+The `slice` function attempts to return a sub-sequence of `sequence_A` that starts
+at `left-position` (inclusive), and ends at `right-position`
+(non-inclusive).
+
```

Comment:
For `slice` there is an alternative to indices that is a better fit for Clarity: Make the first number the offset (how many items to skip/drop) and the second number how many items to take/keep for the result sequence:

(slice "blockstack" u5 u5) ;; Returns (some "stack")
(slice (list 1 2 3 4 5) u3 u1) ;; Returns (some (4))

Unlike with indices, if the second number is a constant, the max length of the result sequence can be trivially determined also when the first numerical argument is a variable. This is a significant benefit for cost estimation in Clarity.

The SIP should include this parameterization as an alternative to indices.

Reply:
The cost estimation would be the same either way.  Either approach requires the implementation do to some bounds-checking to determine whether or not to evaluate to `none` or `some`, and in the latter case, the implementation must calculate the range of elements in the subsequence to consider.

> The SIP should consider this parameterization as an alternative to indices.

I think that regardless of which parameterization is used, the user could easily create a private function that implements the other parameterization if so desired.

Reply:
> The cost estimation would be the same either way.

No. Using `offset` and `length` as parameters makes it trivial to determine the max range of the result sequence:

```clarity
(slice "abcdefgh" i u3)  ;; max length of the return sequence can be estimated to be 3
```
Compare to the equivalent call using indices, where it is not trivial to determine the length of the return sequence:
```clarity
(slice "abcdefgh" i (+ i u3))  ;; max length of the return sequence can be estimated to be 8
```
Thus, `offset` and `length` are a better parameter choice.

Reply:
In both interfaces, the Clarity VM does not know in advance what the length of the resulting slice will be, because both arguments may be expressions whose evaluated values will not be known until runtime.  The length of the sequence that `slice` evaluates to has the same length as `sequence_A` in both cases; and in both cases, the caller will need to use `as-max-len?` to assert a smaller length if that is desired.

Reply:
>  the Clarity VM does not know in advance what the length of the resulting slice will be

That's a failure in the implementation. If the `slice` function takes `offset` and `length` as parameters instead of indices, the Clarity VM should be able to determine in advance that the result of `(slice "abcdefgh" i u3)` is at most 3 characters.

Reply:
> That's a failure in the implementation. If the slice function takes offset and length as parameters instead of indices, the Clarity VM should be able to determine in advance that the result of
(slice "abcdefgh" i u3) is at most 3 characters.

No, it will not.  Consider the case where the `length` parameter is the last 16 bits of the last block's VRF seed.  This value _cannot be known in advance_.

Reply:
>  Consider the case where the length parameter is the last 16 bits of the last block's VRF seed. 

The premise is that with offset and length as arguments, _if the second number (the length) is a constant_, the max length of the result sequence can be trivially determined also when the first numerical argument (the offset) is a variable.

Reply:
A version of `slice` that _must_ take a literal to represent the range of data to be extracted is a non-starter for me, so I guess this argument is a moot point.

Reply:
> A version of slice that must take a literal

This is not suggested.

Reply:
> The premise is that with offset and length as arguments, if the second number (the length) is a constant
>> A version of slice that must take a literal
> This is not suggested.

Either the argument for `length` is permitted to be derived from information that is only known at runtime, or it isn't.  If it isn't, then `slice` cannot be used for its intended use-case (i.e. decoding data uploaded to a smart contract, which may require extracting variable-length fields).  If it's the former, then there's no cost difference between a an offset/length approach versus a start/end approach.  Please help me understand.

Reply:
I had to find extra time to think it through and In this case I have to agree with @jcnelson .
It doesn't really matter whether we'll have `(slice sequenceX start-pos end-pos)` or `(slice sequenceX offset length)` as type of its output **must be** always `(opional sequenceX)` It means that we don't calculate max length of the output at all. It is always the same as the length of the input sequence.

At first glance it feels counterintuitive, but we have to apply to this function exactly the same rule which is applied to all other built-in and user-defined functions. 
> Each function must return value of a known and static data type.

One of the best example for this rule is `concat`. It takes 2 arbitrary long sequences of the same type and returns a sequence of exactly the same type as its inputs and length that is sum of lengths of input sequences.
This simplifies contract validation as it can be performed only once (during deployment).

Example:
```clarity
;; returns (buff 9) because 3 + 3 + 3 = 9
(define-read-only (concat3 (x (buff 3)) (y (buff 3)) (z (buff 3)))
  (concat (concat x y) z)
)

(define-private (test1 (x (buff 20)))
  (print x)
)

(define-private (test2 (x (buff 9)))
  (print x)
)

(define-private (test3 (x (buff 8)))
  (print x)
)

(test1 (concat3 0x 0x 0x)) ;; (buff 9) fits into (buff 20) therefore it prints 0x
(test2 (concat3 0x 0x 0x)) ;; (buff 9) = (buff 9) so it prints 0x
(test3 (concat3 0x 0x 0x)) ;; despite the fact that concat3 returns 0x, type of this value is (buff 9)
                           ;; (buff 9) is longer than (buff 8) therefore we'll get an error at this line
```

And the same example but for slice:
```clarity
;; returns (optional (buff 300))
(define-read-only (cut (input (buff 300)) (x uint) (y uint))
  (slice cut x y)
)

(define-private (test1 (x (optional (buff 400))))
  (print x)
)

(define-private (test2 (x (optional (buff 300))))
  (print x)
)

(define-private (test3 (x (optional (buff 100))))
  (print x)
)

(test1 (cut 0x00010203 u0 u2)) ;; (buff 300) fits into (buff 400) so we can print (some 0x0001)
(test1 (cut 0x00010203 u0 u2)) ;; (buff 300) = (buff 300) - we can print (some 0x0001)
(test1 (cut 0x00010203 u0 u2)) ;; cut returns (some 0x0001) but type of this value is
                               ;; (optional (buff 300)). Because of that we'll get an error at this line

```